### PR TITLE
RFC Testing our work

### PR DESCRIPTION
**Pull request by tahb**
We don't yet have a RFC on how we test our work.

This proposal seeks to agree some higher level principles. The intention is that this forms a basis for which the [standards for testing](https://docs.google.com/document/d/1pSZkrviFAiR-0U-PGixOyC2MfiBdCIWIN66YyLa23nc/edit#) work can sit within. That standard should contain the finer details for how we follow these principles in the way we write tests and the tools we use.

### CODE REVIEW

**Comment by tahb** [comment]
@Gweaton Would you be happy to give this a review first, checking to see I've not missed anything glaring please? Then remove it from draft and notify the wider team to review it. 

If anyone happens to stumble across this in the mean time then by all meanas jump in!

File: rfc-173-testing-our-work.md [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+  and tooling that provisions real infrastructure
+- We MUST write appropriate test coverage for the code we write, inline with our
+  standards for testing. This includes using a combination of feature, request
+  and unit tests. It will also involve setting testing boundaries between
+  systems and third-party software
+- We MUST configure our continuous integration tools to automatically run as
+  deployment pipelines. It MUST prevent deployments completing should any test
+  fail
+- We MUST set up new projects to require 100% code coverage
+
```
Comment (Gweaton) [inline rfc-173-testing-our-work.md:54]: Should we add a few `MUST`s about accessibility testing here? Calum left a few helpful comments in the standards for testing document.

File: rfc-173-testing-our-work.md [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+- We SHOULD write our tests in a consistent and readable way that follows our
+  conventions and patterns set out in our standards for testing
+- We SHOULD use consistent tooling for testing where possible
+- We SHOULD write architecture decision records (ADR) within each project that
+  chooses to diverge from our tools or standards for testing diverge from dxw
+  testing standards
+- We SHOULD aim to increase the minimum test coverage on projects that don't
+  already require 100% as high as we can
+
```
Comment (Gweaton) [inline rfc-173-testing-our-work.md:63]: Is it worth adding a couple of `SHOULD`s here about browser testing e.g if there's client side javascript or for cross-browser testing as we've got in the standards for testing document?

File: rfc-173-testing-our-work.md [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+we achieve them. We propose to document those finer details in our
+[standards for testing](https://docs.google.com/document/d/1pSZkrviFAiR-0U-PGixOyC2MfiBdCIWIN66YyLa23nc/edit#heading=h.owf78h5xtccv)
+that may change more frequently. These standards may eventually become a guide
+in [the dxw Playbook](https://github.com/dxw/playbook).
+
+- We MUST write automated tests for all the code we write that will (or could)
+  serve people in live environments – this includes internal tools for dxw staff
+  and tooling that provisions real infrastructure
+- We MUST write appropriate test coverage for the code we write, inline with our
+  standards for testing. This includes using a combination of feature, request
```
Comment (Gweaton) [inline rfc-173-testing-our-work.md:47]: Worth adding a definition of what `feature, request and unit tests` mean here, or is that best left to the standards for testing document?

File: rfc-173-testing-our-work.md [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+  and unit tests. It will also involve setting testing boundaries between
+  systems and third-party software
+- We MUST configure our continuous integration tools to automatically run as
+  deployment pipelines. It MUST prevent deployments completing should any test
+  fail
+- We MUST set up new projects to require 100% code coverage
```
Comment (pezholio) [inline rfc-173-testing-our-work.md:53]: I think 100% code coverage is pretty much impossible in many cases. There will always be some lines of code that are going to be hard / impossible to test. I think 98% might be more realistic.

Comment (tahb) [reply rfc-173-testing-our-work.md:53]: I think in those cases we should explicitly remove them from the coverage with `# :nocov:` comments rather than allowing a couple of degrees of untestable padding?

Comment (pezholio) [reply rfc-173-testing-our-work.md:53]: Yeah, that's fair. Might be worth codifying that in that case 👍

File: rfc-173-testing-our-work.md (hunk 1 above)
Comment (pezholio) [reply rfc-173-testing-our-work.md:54]: I only have experience using it with Cypress, but I think it'd be good to see if we could use it with Rspec feature tests

File: rfc-173-testing-our-work.md (hunk 3 above)
Comment (tahb) [reply rfc-173-testing-our-work.md:47]: Good point. I'll try and introduce these terms by going back to 'unit' and 'integration' tests and leave it to the standard to define the different types of integration.

File: rfc-173-testing-our-work.md (hunk 2 above)
Comment (tahb) [reply rfc-173-testing-our-work.md:63]: Added browser and OS as a SHOULD 👍 

Testing without javascript is possibly a bit tricker because it's client dependant. We definitely should with apps following the service standard otherwise it'll fail but we can be involved in building JS apps where that standard doesn't apply? I'll add it with a caveat.

File: rfc-173-testing-our-work.md (hunk 1 above)
Comment (tahb) [reply rfc-173-testing-our-work.md:54]: Coming back to this I think there's two parts to it: continous testing through automation and performing an audit. I think the first is a SHOULD. We might be ready to commit to the second becoming a MUST?

**Review by rjw1 — APPROVED** [review]

**Review by Gweaton — APPROVED** [review]

File: rfc-173-testing-our-work.md [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+  be used by people in live environments – this includes internal dxw tools and
+  tooling that provisions real infrastructure
+- We MUST write appropriate test coverage for the code we write inline with our
+  standards for testing. This includes using
+  [the right type of test](https://thoughtbot.com/blog/rails-test-types-and-the-testing-pyramid)
+  as well as setting up and working with appropriate testing boundaries
+- We MUST configure our continuous integration tools to automatically run our
+  tests as part of our deployment pipelines. It MUST prevent deployments
+  completing should any test fail
+- We MUST set up new projects to automatically require 100% code coverage
```
Comment (james) [inline rfc-173-testing-our-work.md:56]: Do we/should we *require* 100% code coverage? Code coverage tools can be helpful guidance/nudge, but I don't think this should be a must.

One reason is I'm keen to ensure that we don't have RFCs that already don't apply to our existing projects - does everything we own have 100% code coverage? Are we going to go back and fix those that don't?

Comment (tahb) [reply rfc-173-testing-our-work.md:56]: 👋 To my knowledge we have never strictly required 100% test coverage at dxw. We have tended to write lots of tests and then sometimes added monitoring for it later and were then unable to get it back to 100%.

Here I'm proposing that only _new_ projects aim for 100% allowing old ones to stay as they are. I quite agree that going back to apply this change over past work isn't feasible. 

Code coverage feedback is a useful as a nudge but from what I've seen there are problems that undermine it when it's not set to 100%:

Untested code gets through CI and slowly builds up quietly until it trips the threshold. The team then face a coverage report that says there are x number of LOC that have no coverage. This puts the unlucky person in a difficult position of having to retrospectively make it 100% or perhaps just fix one or two tests to make CI happy for the time being. The cycle then continues.

I've seen thresholds removed from 96 to 95% to get around this because 96 is totally arbitrary and doesn't mean anything.

I'm know I'm being bold by suggesting it become a MUST rather than a SHOULD but I think it's the right thing to do going forwards? If something can't be tested for a good reason the nudge could be to comment it out with nodoc or similar.

Comment (tahb) [reply rfc-173-testing-our-work.md:56]: Another way might be to have it set to 100% but not enforcing the test suite to hit it. My gut tells me the team would eventually become numb to the noisey nudges?

Comment (jdudley1123) [reply rfc-173-testing-our-work.md:56]: I agree @tahb, I think set 100% as the default for new projects, and if a project has a good reason to change that then they can.

File: rfc-173-testing-our-work.md [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+the details of how we achieve them. We propose to document those finer details
+in our
+[standards for testing](https://docs.google.com/document/d/1pSZkrviFAiR-0U-PGixOyC2MfiBdCIWIN66YyLa23nc/edit#heading=h.owf78h5xtccv)
+that may change more frequently. These standards may eventually become a guide
+in [the dxw Playbook](https://github.com/dxw/playbook).
+
+- We MUST write automated tests for all the code we write that will (or could)
+  be used by people in live environments – this includes internal dxw tools and
+  tooling that provisions real infrastructure
+- We MUST write appropriate test coverage for the code we write inline with our
```
Comment (jdudley1123) [inline rfc-173-testing-our-work.md:49]: Extremely minor point: 'inline' should be 'in line'

**Comment by tahb** [comment]
Can't work out why CI has started to fail due to dead links in another RFC that aren't actually dead!? [I've proposed removing that plugin](https://github.com/dxw/tech-team-rfcs/pull/177) as if it's giving false positives then what can we do?

File: rfc-173-testing-our-work.md (hunk 5 above)
Comment (jacksonj04) [reply rfc-173-testing-our-work.md:56]: Is it worth adding a change coverage requirement for old projects, eg "we SHOULD set up existing projects to require 100% coverage of code changes"? That way we aren't failing on things which are untested already, but we make sure new things get the attention.

Comment (james) [reply rfc-173-testing-our-work.md:56]: I'm happy with what we seem to have agreed on in these comments, which is that new projects should aim to have 100% test coverage unless they have explicit reasons not to. Still, I don't think that's what this will be read as without this chat thread and would be more comfortable with either more qualifiers or as a SHOULD as @jacksonj04 suggested.

Comment (tahb) [reply rfc-173-testing-our-work.md:56]: I'm putting "new" in bold and adding a SHOULD as suggested. Thanks for your contributions!

> - We MUST set up **new** projects to automatically require 100% code coverage
> - We SHOULD set up existing projects to require 100% code coverage
